datarobot_batch_scoring/batch_scoring.py

```diff
@@ -41,53 +45,126 @@ class ShelveError(Exception):
 Batch = collections.namedtuple('Batch', 'id fieldnames data rty_cnt')
 Prediction = collections.namedtuple('Prediction', 'fieldnames data')
 
+SENTINEL = Batch(-1, None, '', -1)
+
 
 class TargetType(object):
     REGRESSION = 'Regression'
     BINARY = 'Binary'
 
 
+def fast_to_csv_chunk(data, header):
+    """Fast routine to format data for prediction api.
+
+    Returns data in unicode.
+    """
+    header = ','.join(header)
+    return ''.join(chain((header, '\n'), data))
+
+
+def slow_to_csv_chunk(data, header):
+    """Slow routine to format data for prediction api.
+
+    Returns data in unicode.
+    """
+    if six.PY3:
+        buf = io.StringIO()
+    else:
+        buf = io.BytesIO()
+
+    writer = csv.writer(buf)
+    writer.writerow(header)
+    writer.writerows(data)
+    return buf.getvalue()
+
+
+class FastReader(object):
+    """A reader that only reads the file in text mode but not parses it. """
+
+    def __init__(self, fd, sep=",", dialect=None, encoding='utf-8'):
+        self.fd = fd
+        reader = csv.reader(self.fd, dialect, delimiter=sep)
+        self.header = next(reader)
+        self.fieldnames = [c.strip() for c in self.header]
+
+    def __iter__(self):
+        self.fd.seek(0)
+        it = iter(self.fd)
+        next(it)  # skip header
+        return it
+
+
+class SlowReader(object):
+    """The slow reader does actual CSV parsing.
+    It supports multiline csv and can be a factor of 50 slower. """
+
+    def __init__(self, fd, sep=",", dialect=None, encoding='utf-8'):
+        self.fd = fd
+        self.sep = sep
+        self.dialect = dialect
+        self.reader = csv.reader(self.fd, dialect, delimiter=sep)
+        self.header = next(self.reader)
+        self.fieldnames = [c.strip() for c in self.header]
+
+    def __iter__(self):
+        self.fd.seek(0)
+        self.reader = csv.reader(self.fd, self.dialect, delimiter=self.sep)
```

Comment:
i think the reader doesnt support a reinitialize -- it will consume lines from the upstream stream (fd) but buffers them internally to support multiline. So if you seek the upstream fd we should reset the reader.